source/concepts/recording.rst

```diff
@@ -25,6 +25,81 @@ Each stream is stored in a separate file.
+  recording and should handle errors accordingly.
+* **the version of the file format**. This version number dictates how the rest of the file will be written or parsed.
+  Any change to the file format needs to increment this file format version. The current version is 2.
+
+A record contains:
+
+* a timestamp encoded as a little endian 16 bytes unsigned integer that indicates the time, in microseconds,
+  since the beginning of the recording.
+  This timestamp indicates the timing of the records and allows synchronisation of a trajectory and a state recording.
+* the size, in bytes, of the record; encoded as an 8 bytes little endian unsigned integer.
+* the record itself as a protobuf message.
+
+In the case of a trajectory recording, each record contains a ``GetFrameResponse`` message.
+This message contains two fields: the frame index and the frame itself.
+The frame index is generally an integer that gets incremented each time the server register a frame to broadcast.
+However, its value is only significant when it is 0 as it means the frame needs to be reset;
+for instance because the server loaded a new simulation. The frame is a :ref:`FrameData <traj-and-frames>`.
+
+In the case of a shared state recording, each record contains a :ref:`StateUpdate <state-updates>` message.
+
+
+Recording with Python
+--------------------------------
+
+How to record
+~~~~~~~~~~~~~
+
+NanoVer sessions can be recorded using the :mod:`nanover.omni.record` module.
+Here is an example of how to define the file names and paths for the recording and pass them to the recording function:
+
+.. code:: python
+
+    from nanover.omni.record import record_from_server
+    # Define the .traj and .state file names and paths
+    traj_path = 'simulation_recording.traj'
+    state_path = 'simulation_recording.state'
+    # create a recording from a server and save it to the files
+    record_from_server("localhost:38801", traj_path, state_path)
+
+How to visualise recordings
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Visualising and playing back recordings can be done using :mod:`nanover.omni.playback` module.
+The Python Server can stream recorded NanoVer streams read by a ``PlaybackSimulation`` object to a client. The client then plays back the recording as if it were
+a live stream.
+The server sends the frames and state updates whilst trying to respect the timing dictated by the timestamps stored
+in the file.
+
+.. code:: python
+
+    from nanover.omni import OmniRunner
+    from nanover.omni.playback import PlaybackSimulation
+    simulation_recording = PlaybackSimulation(name='simulation-recording', traj='files/simulation_recording.traj',
+                                           state='files/simulation_recording.state')
+    # Create a runner for the simulation
+    recording_runner = OmniRunner.with_basic_server(simulation_recording, name='simulation-recording-server')
```

Comment:
Again, split over multiple lines so the code example doesn't run off of the page

Reply:
Done in commit [52dc633](https://github.com/IRL2/nanover-docs/pull/113/commits/52dc633153e8d43308b50314cf888181b9394a23).